___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright by Pierian Data Inc.</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [4]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [5]:
# help(pd.DataFrame)

In [6]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [7]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [8]:
myindex = ['CA','NY','AZ','TX']

In [9]:
mycolumns = ['Jan','Feb','Mar']

In [10]:
df=pd.DataFrame(mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [11]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [12]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [13]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int32
 1   Feb     4 non-null      int32
 2   Mar     4 non-null      int32
dtypes: int32(3)
memory usage: 80.0+ bytes


In [15]:
#represent useful information 


# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [16]:
pwd

'C:\\Users\\PRINCE\\Desktop\\FINANCE\\Algo_trading\\03-Core-Pandas'

#### List the files in your current directory with ls

In [17]:
ls

 Volume in drive C has no label.
 Volume Serial Number is A446-1A61

 Directory of C:\Users\PRINCE\Desktop\FINANCE\Algo_trading\03-Core-Pandas

31-05-2023  00:32    <DIR>          .
27-06-2023  16:24    <DIR>          ..
31-05-2023  00:32    <DIR>          .ipynb_checkpoints
28-09-2022  21:16    <DIR>          __pycache__
07-10-2022  22:52           565,403 00-Series.ipynb
07-10-2022  23:53           211,637 01-DataFrames.ipynb
28-09-2022  21:16           194,591 02-Conditional-Filtering.ipynb
28-09-2022  21:16           196,385 03-Useful-Methods.ipynb
11-10-2022  19:28            65,135 04-Combining-DataFrames.ipynb
11-10-2022  19:56            65,247 05-Inputs-and-Outputs.ipynb
14-10-2022  22:21           148,410 06-Core-Pandas-Finance-Exercise.ipynb
13-10-2022  20:24            77,549 07-Core-Pandas-Finance-Exercise-SOLUTIONS.ipynb
28-09-2022  21:16            18,676 constituents.csv
28-09-2022  21:16            73,270 constituents-financials.csv
28-09-2022  21:16                46 

In [42]:
df = pd.read_csv('C:\\Users\\PRINCE\\Desktop\\FINANCE\\Algo_trading\\03-Core-Pandas\\tips.csv')


In [43]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [65]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID',
       'tip_percentage'],
      dtype='object')

In [66]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [67]:
df.head(3)

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.7    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.5    Male     No  Sun  Dinner     3          7.003333   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734

In [68]:
df.tail(3)

total_bill  tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.0    Male    Yes   Sat  Dinner     2            11.335   
242       17.82  1.8    Male     No   Sat  Dinner     2             8.910   
243       18.78  3.0  Female     No  Thur  Dinner     2             9.390   

          Payer Name         CC Number Payment ID  tip_percentage  
241       Keith Wong  6011891618747196    Sat3880        8.822232  
242     Dennis Dixon     4375220550950      Sat17        9.820426  
243  Michelle Hardin  3511451626698139    Thur672       15.974441

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
 11  tip_percentage    244 non-null    float64
dtypes: float64(4), int64(2), object(6)
memory usage: 23.0+ KB


In [70]:
len(df)

244

In [71]:
df.describe()

total_bill         tip        size  price_per_person     CC Number  \
count  244.000000  244.000000  244.000000         244.00000  2.440000e+02   
mean    19.785943    2.997951    2.569672           7.88823  2.563496e+15   
std      8.902412    1.385060    0.951100           2.91435  2.369340e+15   
min      3.070000    1.000000    1.000000           2.87500  6.040679e+10   
25%     13.347500    2.000000    2.000000           5.80250  3.040731e+13   
50%     17.795000    2.900000    2.000000           7.25500  3.525318e+15   
75%     24.127500    3.600000    3.000000           9.39000  4.553675e+15   
max     50.810000   10.000000    6.000000          20.27500  6.596454e+15   

       tip_percentage  
count      244.000000  
mean        16.080258  
std          6.107220  
min          3.563814  
25%         12.912736  
50%         15.476977  
75%         19.147549  
max         71.034483

In [72]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.997951e+00  1.385060e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888230e+00  2.914350e+00  2.875000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   
tip_percentage    244.0  1.608026e+01  6.107220e+00  3.563814e+00   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.600000e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.802500e+00  7.255000e+00  9.390000e+00  2.027500e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15  
tip_percentage    1.291274e+01  1.547698e+01  1.914755e+01  7.103448e+01

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [73]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.7    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.5    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.3    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.6  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Grab a Single Column

In [74]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [75]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

In [76]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']]

total_bill  tip
0         16.99  1.0
1         10.34  1.7
2         21.01  3.5
3         23.68  3.3
4         24.59  3.6
..          ...  ...
239       29.03  5.9
240       27.18  2.0
241       22.67  2.0
242       17.82  1.8
243       18.78  3.0

[244 rows x 2 columns]

#### Create New Columns

In [77]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [78]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.7    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.5    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.3    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.6  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.885815  
1      Douglas Tucker  4478071379779230    Sun4608       16.441006  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.935811  
4        Tonya Carter  4832732618637221    Sun2251       14.640098

In [79]:
df['price_per_person'] = df['total_bill'] / df['size']

In [80]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.7    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.5    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.3    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.6  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.885815  
1      Douglas Tucker  4478071379779230    Sun4608       16.441006  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.935811  
4        Tonya Carter  4832732618637221    Sun2251       14.640098

In [81]:
df['tip']=np.round(df['tip'],1)

In [82]:
df

total_bill  tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.0  Female     No   Sun  Dinner     2          8.495000   
1         10.34  1.7    Male     No   Sun  Dinner     3          3.446667   
2         21.01  3.5    Male     No   Sun  Dinner     3          7.003333   
3         23.68  3.3    Male     No   Sun  Dinner     2         11.840000   
4         24.59  3.6  Female     No   Sun  Dinner     4          6.147500   
..          ...  ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.9    Male     No   Sat  Dinner     3          9.676667   
240       27.18  2.0  Female    Yes   Sat  Dinner     2         13.590000   
241       22.67  2.0    Male    Yes   Sat  Dinner     2         11.335000   
242       17.82  1.8    Male     No   Sat  Dinner     2          8.910000   
243       18.78  3.0  Female     No  Thur  Dinner     2          9.390000   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.885815  
1        Douglas Tucker  4478071379779230    Sun4608       16.441006  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.935811  
4          Tonya Carter  4832732618637221    Sun2251       14.640098  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.323803  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17       10.101010  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [83]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [84]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [85]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.7    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.5    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.3    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.6  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.885815  
1      Douglas Tucker  4478071379779230    Sun4608       16.441006  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.935811  
4        Tonya Carter  4832732618637221    Sun2251       14.640098

#### Remove Columns

In [86]:
# df.drop('tip_percentage',axis=1)

In [87]:
df = df.drop("tip_percentage",axis=1)

In [88]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.7    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.5    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.3    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.6  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [89]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.7    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.5    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.3    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.6  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [90]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [91]:
df.set_index('Payment ID')

total_bill  tip     sex smoker   day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.0  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.7    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.6  Female     No   Sun  Dinner     4   
...                ...  ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.9    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.0  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.0    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.8    Male     No   Sat  Dinner     2   
Thur672          18.78  3.0  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

In [92]:
df.head()

total_bill  tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.0  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.7    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.5    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.3    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.6  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [93]:
df = df.set_index('Payment ID')

In [94]:
df.head()

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun2959          16.99  1.0  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.7    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.6  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [95]:
df = df.reset_index()

In [96]:
df.head()

Payment ID  total_bill  tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.0  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.7    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.5    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.3    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.6  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [97]:
df.head()

Payment ID  total_bill  tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.0  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.7    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.5    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.3    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.6  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

In [98]:
df = df.set_index('Payment ID')

In [99]:
df.head()

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun2959          16.99  1.0  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.7    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.6  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Row

In [100]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                                1.0
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [101]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                                1.0
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [102]:
df.iloc[0:4]

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun2959          16.99  1.0  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.7    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [103]:
df.loc[['Sun2959','Sun5260']]

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun2959          16.99  1.0  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [104]:
df.head()

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun2959          16.99  1.0  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.7    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.6  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [105]:
df.drop('Sun2959',axis=0).head()

total_bill  tip     sex smoker  day    time  size  \
Payment ID                                                      
Sun4608          10.34  1.7    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.5    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.3    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.6  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.7    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [106]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [107]:
one_row = df.iloc[0]

In [108]:
one_row

total_bill                       16.99
tip                                1.0
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [109]:
type(one_row)

pandas.core.series.Series

In [110]:
df.tail()

total_bill  tip     sex smoker   day    time  size  \
Payment ID                                                       
Sat2657          29.03  5.9    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.0  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.0    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.8    Male     No   Sat  Dinner     2   
Thur672          18.78  3.0  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat2657                 9.68    Michael Avila  5296068606052842  
Sat1766                13.59   Monica Sanders  3506806155565404  
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [111]:
df.append(one_row).tail()

C:\Users\PRINCE\AppData\Local\Temp\ipykernel_9616\112637673.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df.append(one_row).tail()


total_bill  tip     sex smoker   day    time  size  \
Payment ID                                                       
Sat1766          27.18  2.0  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.0    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.8    Male     No   Sat  Dinner     2   
Thur672          18.78  3.0  Female     No  Thur  Dinner     2   
Sun2959          16.99  1.0  Female     No   Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  
Sun2959                 8.49  Christy Cunningham  3560325168603410

--------